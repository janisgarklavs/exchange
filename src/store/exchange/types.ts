export interface Exchange {
  base: string;
  date: string;
  rates: Record<string, number>;
}

export type ApiResponse = Record<string, any>;

export enum ExchangeActionTypes {
  FETCH_REQUEST = "@@exchange/FETCH_REQUEST",
  FETCH_SUCCESS = "@@exchange/FETCH_SUCCESS",
  FETCH_ERROR = "@@exchange/FETCH_ERROR"
}

export interface ExchangeState {
  readonly loading: boolean;
  readonly data?: Exchange;
  readonly errors?: string;
}
