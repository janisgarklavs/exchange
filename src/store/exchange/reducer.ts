import { Reducer } from "redux";
import { ExchangeState, ExchangeActionTypes } from "./types";

const initialState: ExchangeState = {
  data: undefined,
  errors: undefined,
  loading: false
};

const reducer: Reducer<ExchangeState> = (state = initialState, action) => {
  switch (action.type) {
    case ExchangeActionTypes.FETCH_REQUEST: {
      return { ...state, loading: true };
    }
    case ExchangeActionTypes.FETCH_SUCCESS: {
      return {
        ...state,
        loading: false,
        data: action.payload,
        errors: ""
      };
    }
    case ExchangeActionTypes.FETCH_ERROR: {
      return {
        ...state,
        loading: false,
        errors: action.payload
      };
    }
    default: {
      return state;
    }
  }
};

export { reducer as exchangeReducer };
