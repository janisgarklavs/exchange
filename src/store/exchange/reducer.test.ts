import { exchangeReducer } from "./reducer";
import { Exchange, ExchangeActionTypes } from "./types";

describe("exchange reducer", () => {
  describe("INITIAL_STATE", () => {
    test("is correct", () => {
      const action = { type: "dummy_action" };
      expect(exchangeReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe(ExchangeActionTypes.FETCH_REQUEST, () => {
    test("returns the correct state", () => {
      const action = { type: ExchangeActionTypes.FETCH_REQUEST };
      const expectedState = {
        errors: undefined,
        loading: true,
        data: undefined
      };
      expect(exchangeReducer(undefined, action)).toEqual(expectedState);
    });
  });

  describe(ExchangeActionTypes.FETCH_SUCCESS, () => {
    test("returns the correct state", () => {
      const action = {
        type: ExchangeActionTypes.FETCH_SUCCESS,
        payload: <Exchange>{}
      };
      const expectedState = {
        errors: "",
        loading: false,
        data: <Exchange>{}
      };
      expect(exchangeReducer(undefined, action)).toEqual(expectedState);
    });
  });

  describe(ExchangeActionTypes.FETCH_ERROR, () => {
    test("returns the correct state", () => {
      const action = {
        type: ExchangeActionTypes.FETCH_ERROR,
        payload: "something wrong"
      };
      const expectedState = {
        errors: "something wrong",
        loading: false,
        data: undefined
      };
      expect(exchangeReducer(undefined, action)).toEqual(expectedState);
    });
  });
});
