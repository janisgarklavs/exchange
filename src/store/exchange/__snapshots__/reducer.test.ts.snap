// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exchange reducer INITIAL_STATE is correct 1`] = `
Object {
  "data": undefined,
  "errors": undefined,
  "loading": false,
}
`;
